action.php

```diff
@@ -29,19 +29,103 @@
 switch ( $_POST['op'] ) {
 	case 'validateApiKey':
 		// Get request params.
-		$key = isset( $_POST['key'] ) ? trim( $_POST['key'] ) : '';
-
-		// Validate api key with remote request.
-		$request = new Wp_Sendsmaily_Request(
-			'https://www.sendsmaily.net/validate_key.php',
-			array( 'key' => $key )
+		$subdomain = isset( $_POST['subdomain'] ) ? trim( $_POST['subdomain'] ) : '';
+		$username  = isset( $_POST['username'] ) ? trim( $_POST['username'] ) : '';
+		$password  = isset( $_POST['password'] ) ? trim( $_POST['password'] ) : '';
+
+		$params = array(
+			'subdomain' => $subdomain,
+			'username'  => $username,
+			'password'  => $password,
 		);
-		$result  = $request->exec();
-		$data    = $result['data'];
 
-		// Handle errors.
-		if ( isset( $result['code'] ) && $result['code'] >= 200 ) {
-			$result['error'] = true;
+		// Sanitize fields.
+		foreach ( $params as $key => $value ) {
+			$params[ $key ] = wp_unslash( sanitize_text_field( $value ) );
+		}
+
+		// Normalize subdomain.
+		// First, try to parse as full URL. If that fails, try to parse as subdomain.sendsmaily.net, and
+		// if all else fails, then clean up subdomain and pass as is.
+		if ( filter_var( $params['subdomain'], FILTER_VALIDATE_URL ) ) {
+			$url                 = wp_parse_url( $params['subdomain'] );
+			$parts               = explode( '.', $url['host'] );
+			$params['subdomain'] = count( $parts ) >= 3 ? $parts[0] : '';
+		} elseif ( preg_match( '/^[^\.]+\.sendsmaily\.net$/', $params['subdomain'] ) ) {
+			$parts               = explode( '.', $params['subdomain'] );
+			$params['subdomain'] = $parts[0];
+		}
+
+		// Show error messages to user if no data is entered to form.
+		if ( $params['subdomain'] === '' ) {
+			// Don't refresh the page.
+			$_POST['refresh'] = '';
```

Comment:
Try not to modify `$_POST` variable values. If you need to override refresh state on certain cases then create a variable before `switch` statement that you can modify and use later to determine refresh state.

Reply:
Added `$refresh_blocked` variable in https://github.com/sendsmaily/sendsmaily-wordpress-plugin/pull/11/commits/2f12e97d5d9455b790d44ac16c15bfd90ed332db

Reply:
A simpler solution (without introducing new variable) is to move `$refresh` variable declaration from line 317 to 28:
```php
$refresh = ( isset( $_POST[ 'refresh' ] ) && $_POST['refresh'] == 1 );
```
and set `$refresh = false;` where needed (instead of `$refresh_blocked = true;`).

Reply:
Moved `$refresh` to top https://github.com/sendsmaily/sendsmaily-wordpress-plugin/pull/11/commits/135e81939a9e157af3afd10860ae9872e97583d1